File: server/internal/auth/guestname.go
package auth

import (
	"hash/fnv"
	"math/rand"
)

// The name a player is given when they have not said what to call themselves.
//
// Every guest used to be "Player" (the app's prefilled field) or "Guest" (the
// SSH host, and any client that sent no name at all), which at a table of four
// is not a name but a shrug: the seat tiles read Player, Player, Player, and
// the only way to tell whose turn it is is to count. A name has to *identify*
// before it does anything else.
//
// Numbering would identify — Guest 1, Guest 2 — and is rejected for the reason
// written over suggestUsername: a number is a queue ticket, and a queue ticket
// is a poor thing to be called for an evening of cards. So the default is a
// two-word name drawn from a roster, in the same spirit as the bot personas in
// module.Persona (Rookie Rita, Master Miroslav) but deliberately a different
// *shape* — those are adjective-plus-forename, these are adjective-plus-
// creature, so a guest is never mistaken for an opponent nobody is sitting at.
//
// The roster is English and untranslated, exactly as the personas and the
// avatar labels are. That is a considered position rather than an oversight:
// these are proper names, the thing a translation bundle is least able to
// help with, and the field is editable on the very screen that suggests one.
//
// A near-copy of this roster lives in the app, in
// client-react-native/src/lib/guestName.ts, so the name can be shown in the
// field before anything is sent. The two are allowed to drift. Nothing ever
// compares them — unlike the message keys in serverKeys.json, where drift
// reaches a player as SCREAMING_SNAKE — and a guest named from either list is
// equally well named.
//
// 64 x 64 is 4096 names, which puts two guests at a six-seat table sharing one
// at about one table in 270. That is rare enough to be a curiosity rather than
// a defect, and the remedy — typing over the suggestion — is in front of them.
var guestAdjectives = []string{
	"Amber", "Arctic", "Autumn", "Bold", "Brave", "Brisk", "Bright", "Calm",
	"Canny", "Cheery", "Clever", "Copper", "Coral", "Crimson", "Curious", "Daring",
	"Dusty", "Eager", "Emerald", "Fearless", "Fleet", "Frosty", "Gentle", "Golden",
	"Grand", "Happy", "Hasty", "Honest", "Humble", "Indigo", "Ivory", "Jolly",
	"Keen", "Kindly", "Lively", "Lucky", "Merry", "Mighty", "Nimble", "Noble",
	"Patient", "Plucky", "Polite", "Proud", "Quick", "Quiet", "Rapid", "Restless",
	"Royal", "Ruby", "Rusty", "Scarlet", "Sharp", "Silver", "Sly", "Snowy",
	"Solid", "Spry", "Steady", "Stormy", "Sunny", "Swift", "Velvet", "Witty",
}

var guestNouns = []string{
	"Otter", "Badger", "Falcon", "Heron", "Marten", "Lynx", "Sparrow", "Magpie",
	"Raven", "Robin", "Fox", "Hare", "Stag", "Ibex", "Bison", "Boar",
	"Wolf", "Owl", "Crane", "Swallow", "Finch", "Kestrel", "Osprey", "Puffin",
	"Curlew", "Beaver", "Weasel", "Ferret", "Mole", "Dormouse", "Squirrel", "Hedgehog",
	"Chamois", "Lark", "Wren", "Starling", "Kite", "Harrier", "Pike", "Perch",
	"Trout", "Salmon", "Newt", "Frog", "Turtle", "Bee", "Moth", "Cricket",
	"Beetle", "Firefly", "Dragonfly", "Comet", "Ember", "Lantern", "Compass", "Anchor",
	"Beacon", "Pebble", "Willow", "Aspen", "Birch", "Juniper", "Thistle", "Clover",
}

// GuestNameFor invents a display name, the same one every time for the same
// seed.
//
// Stability is the point of the seed, and the seed callers pass is the guest
// id — the device's durable identity. A guest who signs out and comes back, or
// who plays from a terminal that stores no name of its own, is greeted as the
// same person rather than as somebody new each time; and because the id is
// already unique per device, two devices agreeing on a name is a matter of
// chance rather than of everybody landing on the same default.
//
// An empty seed means there is nothing durable to hang a name on yet, so one
// is drawn at random. That is still 4096 ways to not be called Player.
func GuestNameFor(seed string) string {
	if seed == "" {
		return guestName(rand.Uint64())
	}
	h := fnv.New64a()
	_, _ = h.Write([]byte(seed))
	return guestName(h.Sum64())
}

// guestName cuts one number into the two independent indices. Dividing out the
// first list's length before taking the second index keeps the halves from
// moving together, which is what a pair of plain moduli of the same number
// would do whenever the two lengths share a factor — and these two are equal.
func guestName(n uint64) string {
	adj := guestAdjectives[n%uint64(len(guestAdjectives))]
	noun := guestNouns[(n/uint64(len(guestAdjectives)))%uint64(len(guestNouns))]
	return adj + " " + noun
}

File: client-react-native/src/lib/guestName.ts
/**
 * The name offered to a guest who has not said what to call themselves.
 *
 * The field on the guest screen used to open with "Player" in it, and most
 * people leave a prefilled field alone — which is how a table ends up seating
 * Player, Player and Player, and why the only way to follow whose turn it is
 * becomes counting seats. A default name has to identify before it is anything
 * else.
 *
 * Numbering would identify — Player 1, Player 2 — and is not what we do: a
 * number is a queue ticket, a poor thing to be called for an evening of cards.
 * So a name is drawn from a roster instead, in the same spirit as the bot
 * personas the server seats (Rookie Rita, Master Miroslav) but deliberately a
 * different shape — those are adjective-plus-forename, these are
 * adjective-plus-creature — so a guest is never taken for an opponent nobody
 * is sitting at.
 *
 * The roster is English and not translated, as the avatar labels in
 * `components/avatars/catalogue.ts` are not. These are proper names, the thing
 * a locale bundle helps with least, and the suggestion is sitting in an
 * editable field on the screen that makes it.
 *
 * The server keeps a near-copy of this roster (`server/internal/auth/
 * guestname.go`) for clients that send no name at all — the SSH host, mainly.
 * The two lists are allowed to drift: nothing compares them, and a guest named
 * from either is equally well named. That is the opposite of the rule for
 * `serverKeys.json`, where drift reaches a player as SCREAMING_SNAKE.
 */

const ADJECTIVES = [
  'Amber', 'Arctic', 'Autumn', 'Bold', 'Brave', 'Brisk', 'Bright', 'Calm',
  'Canny', 'Cheery', 'Clever', 'Copper', 'Coral', 'Crimson', 'Curious', 'Daring',
  'Dusty', 'Eager', 'Emerald', 'Fearless', 'Fleet', 'Frosty', 'Gentle', 'Golden',
  'Grand', 'Happy', 'Hasty', 'Honest', 'Humble', 'Indigo', 'Ivory', 'Jolly',
  'Keen', 'Kindly', 'Lively', 'Lucky', 'Merry', 'Mighty', 'Nimble', 'Noble',
  'Patient', 'Plucky', 'Polite', 'Proud', 'Quick', 'Quiet', 'Rapid', 'Restless',
  'Royal', 'Ruby', 'Rusty', 'Scarlet', 'Sharp', 'Silver', 'Sly', 'Snowy',
  'Solid', 'Spry', 'Steady', 'Stormy', 'Sunny', 'Swift', 'Velvet', 'Witty',
] as const;

const NOUNS = [
  'Otter', 'Badger', 'Falcon', 'Heron', 'Marten', 'Lynx', 'Sparrow', 'Magpie',
  'Raven', 'Robin', 'Fox', 'Hare', 'Stag', 'Ibex', 'Bison', 'Boar',
  'Wolf', 'Owl', 'Crane', 'Swallow', 'Finch', 'Kestrel', 'Osprey', 'Puffin',
  'Curlew', 'Beaver', 'Weasel', 'Ferret', 'Mole', 'Dormouse', 'Squirrel', 'Hedgehog',
  'Chamois', 'Lark', 'Wren', 'Starling', 'Kite', 'Harrier', 'Pike', 'Perch',
  'Trout', 'Salmon', 'Newt', 'Frog', 'Turtle', 'Bee', 'Moth', 'Cricket',
  'Beetle', 'Firefly', 'Dragonfly', 'Comet', 'Ember', 'Lantern', 'Compass', 'Anchor',
  'Beacon', 'Pebble', 'Willow', 'Aspen', 'Birch', 'Juniper', 'Thistle', 'Clover',
] as const;

/** How many names the roster can make. 64 x 64, and the reason no suffix is
 *  needed: two guests at a six-seat table collide about one table in 270. */
export const GUEST_NAME_COUNT = ADJECTIVES.length * NOUNS.length;

/**
 * A number from a string. The same idiom `avatarFor` hashes an id with, kept
 * identical on purpose — a face and a name derived from one id by one method
 * are two views of the same coin toss rather than two unrelated ones.
 */
function hash(s: string): number {
  let h = 0;
  for (let i = 0; i < s.length; i++) h = (h * 31 + s.charCodeAt(i)) >>> 0;
  return h;
}

/**
 * The name to offer.
 *
 * Given this device's guest id it is the *same* name every time, which is what
 * a returning guest wants: the person who played as Nimble Heron last week and
 * signed out is offered Nimble Heron again, not a stranger's name. With no id
 * — a first run, before the server has issued one — one is drawn at random,
 * which is still 4096 ways to not be called Player.
 *
 * Dividing out the first list's length before the second index keeps the two
 * halves independent; two plain moduli of the same number would move together,
 * because the lengths are equal.
 */
export function guestNameFor(seed?: string | null): string {
  const n = seed ? hash(seed) : Math.floor(Math.random() * GUEST_NAME_COUNT);
  const adjective = ADJECTIVES[n % ADJECTIVES.length]!;
  const noun = NOUNS[Math.floor(n / ADJECTIVES.length) % NOUNS.length]!;
  return `${adjective} ${noun}`;
}
